src/Client/ClientBase.cpp

```diff
@@ -2270,16 +2241,26 @@ void ClientBase::processParsedSingleQuery(const String & full_query, const Strin
                     "Processing async inserts with both inlined and external data (from stdin or infile) is not supported");
         }
 
+        String query;
+        /// An INSERT query may have the data that follows query text.
+        /// Send part of the query without data, because data will be sent separately.
+        /// But for asynchronous inserts we don't extract data, because it's needed
+        /// to be done on server side in that case (for coalescing the data from multiple inserts on server side).
```

Comment:
If we send data separately anyway, will insert become synchronous, or will async insert just use more cpu on the server (on squashing the blocks?)? Would tests notice if this is broken and e.g. `is_async_insert_with_inlined_data` is always false?

Reply:
>If we send data separately anyway, will insert become synchronous, or will async insert just use more cpu on the server (on squashing the blocks?)?

The only benefit of async INSERT over TCP AFAICS is decreasing number of real INSERTs i.e. number of parts for MergeTree family

> Would tests notice if this is broken and e.g. is_async_insert_with_inlined_data is always false?

They should, since all of the combinations are checked in the test.